src/public/js/rs.js

```diff
@@ -21,6 +105,94 @@ var RS = (function() {
     }, '');
   }
 
+  // Encodes using systematic BCH Reed-Solomon on Galois Field 256
+  // Message should be passed in as length k block
+  // k specifies message length
+  // n specifies block length (number of symbols transmitted)
+  // n - k is the number of code symbols added.
+  // Encoding procedure has the largest order coefficient on the left
+  // and the smallest on the right. IE: [1,2,3] -> x^2 + 2x + 3
+  function encodeRSBlock(msg, n, k) {
+    remainder = polynomialDiv(genPoly, msg, n, k)
+    if (remainder == null) {
+      return null;
+    }
+    codedMsg = new Uint8Array(n)
+    for (i = 0; i < k; i++) {
+      codedMsg[i] = msg[i];
+    }
+    for (i = k; i < n; i++) {
+      codedMsg[i] = remainder[i - k]
+    }
+  }
+  function decodeRSBlock(codedMsg, n, k) {
+    remainder = polynomialDiv(genPoly, codedMsg, n, k)
+    if (remainder == null) {
+      return null;
+    }
+    errors = false;
+    for (i = 0; i < n-k; i++) {
+      if (remainder[i] != 0) {
+        errors = true
+      }
+    }
+    if (errors) {
+      // TODO ERROR CORRECTION
+      return null
+    } else {
+      msg = new Uint8Array(k)
+      for (i = 0; i < k; i++) {
+        msg[i] = codedMsg[i]
+      }
+      return msg
+    }
+  }
+
+  function polynomialDiv(divisor, dividend, n, k) {
+    if (divisor.length != k || dividend.length != n) {
+      return null;
+    }
+    // precompute divisors highest order inverse and distribute to all of divisor
+    invVal = multInv[divisor[0]];
+    for (i = 0; i < k; i++){
+      divisor[i] = fieldMultiply(divisor[i], invVal);
+    }
+    for (i = 0; i < n-k; i++) {
+      // find how many of highest order of scaled divisor go in to
+      // current highest order of leftover dividend
+      // Because of precompute it is simply the coefficient of the highest order of dividend
+      quotient = dividend[i];
+
+      // find scaled divisor to subtract from current remainder of dividend
+      // by multiplying each coefficient of precomputed divisor by current
+      // highest order quotient
+      scaledDivisor = new Uint8Array(k);
+      for (j = 0; j < k; j++) {
+        scaledDivisor[j] = fieldMultiply(divisor[j], quotient);
+      }
+
+      // Subtract scaled divisor from current remainder of dividend
+      // to get new remainder of dividend which will be 1 order less
+      for (j = i; j < i+k; j++) {
+        // Galois field subtraction is xor with base field 2
```

Comment:
(this is me being confused about whether all Galois fields are extension fields)